tests/test_base_api_client.py

```diff
@@ -33,29 +36,70 @@ def _empty_context_manager():
 
 
 class TestBaseApiClient(object):
-    def test_connection_error_raises_api_error(self, base_client, raw_rmock):
-        raw_rmock.side_effect = requests.exceptions.ConnectionError(
-            None
-        )
-        with pytest.raises(HTTPError) as e:
-            base_client._request("GET", '/')
+    @mock.patch('urllib3.connectionpool.HTTPConnectionPool._make_request')
+    @mock.patch('dmapiclient.base.BaseAPIClient.RETRIES_BACKOFF_FACTOR', 0)
+    def test_client_retries_on_connection_error_and_raises_api_error(self, _make_request, base_client):
+        _make_request.side_effect = NewConnectionError(mock.Mock(), 'Im a message')
+        retry_count = randint(1, 3)
+
+        with mock.patch('dmapiclient.base.BaseAPIClient.RETRIES', retry_count):
+            with pytest.raises(HTTPError) as e:
+                base_client._request("GET", '/')
 
-        assert e.value.message == 'None\nConnectionError(None,)'
+        requests = _make_request.call_args_list
+
+        assert len(requests) == retry_count + 1
+        assert all((request[0][1], request[0][2]) == ('GET', '/') for request in requests)
+
+        assert 'ConnectionError' in e.value.message
         assert e.value.status_code == REQUEST_ERROR_STATUS_CODE
 
-    def test_http_error_raises_api_error(self, base_client, rmock):
-        rmock.request(
-            "GET",
-            "http://baseurl/",
-            text="Internal Error",
-            status_code=500)
+    @mock.patch('urllib3.connectionpool.HTTPConnectionPool._make_request')
+    @mock.patch('dmapiclient.base.BaseAPIClient.RETRIES_BACKOFF_FACTOR', 0)
+    def test_client_retries_on_read_error_and_raises_api_error(self, _make_request, base_client):
+        _make_request.side_effect = ProtocolError(mock.Mock(), 'Im a message')
+        retry_count = randint(1, 3)
```

Comment:
Rather than introducing randomness & possibly random failures, you could just parametrize across all these values.

Reply:
Yeah, i thought this might be nice way of not having to run the test multiple times each test run though. What possible random failures do you foresee?

Reply:
Well just that if someone breaks the 3-retry case it will only fail one in three times -> could make it thru travis, then start giving people mysterious test failures. Cue much "what have I done??" head scratching.

Reply:
(that said, I'm also guilty of using random in tests, but I'm considering removing it sometime)